File: server/src/main/java/dev/ilionx/workshop/api/appointment/controller/AppointmentController.java
package dev.ilionx.workshop.api.appointment.controller;

import dev.ilionx.workshop.api.appointment.model.Appointment;
import dev.ilionx.workshop.api.appointment.model.mapper.AppointmentMapper;
import dev.ilionx.workshop.api.appointment.model.request.CreateAppointmentRequest;
import dev.ilionx.workshop.api.appointment.model.request.UpdateAppointmentRequest;
import dev.ilionx.workshop.api.appointment.model.response.AppointmentResponse;
import dev.ilionx.workshop.api.appointment.service.AppointmentService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;

import java.util.List;
import jakarta.validation.Valid;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import static dev.ilionx.workshop.api.Paths.APPOINTMENTS;
import static dev.ilionx.workshop.api.Paths.APPOINTMENT_BY_ID;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.NO_CONTENT;
import static org.springframework.http.HttpStatus.OK;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

/**
 * REST controller for managing appointments.
 */
@Tag(
    name = "Appointment",
    description = "Appointment scheduling and management endpoints"
)
@RestController
@RequiredArgsConstructor
public class AppointmentController {

    private final AppointmentService appointmentService;
    private final AppointmentMapper appointmentMapper;

    @ResponseStatus(OK)
    @Operation(
        summary = "Get all appointments",
        description = "Returns all appointments in the system"
    )
    @GetMapping(
        path = APPOINTMENTS,
        produces = APPLICATION_JSON_VALUE
    )
    public ResponseEntity<List<AppointmentResponse>> getAllAppointments() {
        final List<Appointment> appointments = appointmentService.findAll();
        return ResponseEntity.status(OK).body(appointmentMapper.toResponseList(appointments));
    }

    @ResponseStatus(OK)
    @Operation(
        summary = "Get appointment by ID",
        description = "Returns a single appointment by its ID"
    )
    @GetMapping(
        path = APPOINTMENT_BY_ID,
        produces = APPLICATION_JSON_VALUE
    )
    public ResponseEntity<AppointmentResponse> getAppointmentById(
        @PathVariable final Integer appointmentId
    ) {
        final Appointment appointment = appointmentService.findById(appointmentId);
        return ResponseEntity.status(OK).body(appointmentMapper.toResponse(appointment));
    }

    @ResponseStatus(CREATED)
    @Operation(
        summary = "Create appointment",
        description = "Creates a new appointment for a pet with a vet"
    )
    @PostMapping(
        path = APPOINTMENTS,
        consumes = APPLICATION_JSON_VALUE,
        produces = APPLICATION_JSON_VALUE
    )
    public ResponseEntity<AppointmentResponse> createAppointment(
        @Valid
        @RequestBody final CreateAppointmentRequest request
    ) {
        final Appointment createdAppointment = appointmentService.create(request);
        return ResponseEntity.status(CREATED).body(appointmentMapper.toResponse(createdAppointment));
    }

    @ResponseStatus(OK)
    @Operation(
        summary = "Update appointment",
        description = "Updates an existing appointment"
    )
    @PutMapping(
        path = APPOINTMENT_BY_ID,
        consumes = APPLICATION_JSON_VALUE,
        produces = APPLICATION_JSON_VALUE
    )
    public ResponseEntity<AppointmentResponse> updateAppointment(
        @PathVariable final Integer appointmentId,
        @Valid
        @RequestBody final UpdateAppointmentRequest request
    ) {
        final Appointment appointment = appointmentService.update(appointmentId, request);
        return ResponseEntity.status(OK).body(appointmentMapper.toResponse(appointment));
    }

    @ResponseStatus(OK)
    @Operation(
        summary = "Cancel appointment",
        description = "Cancels an appointment by setting its status to CANCELLED"
    )
    @PatchMapping(
        path = APPOINTMENT_BY_ID + "/cancel",
        produces = APPLICATION_JSON_VALUE
    )
    public ResponseEntity<AppointmentResponse> cancelAppointment(
        @PathVariable final Integer appointmentId
    ) {
        final Appointment appointment = appointmentService.cancel(appointmentId);
        return ResponseEntity.status(OK).body(appointmentMapper.toResponse(appointment));
    }

    @ResponseStatus(OK)
    @Operation(
        summary = "Complete appointment",
        description = "Marks an appointment as completed by setting its status to COMPLETED"
    )
    @PatchMapping(
        path = APPOINTMENT_BY_ID + "/complete",
        produces = APPLICATION_JSON_VALUE
    )
    public ResponseEntity<AppointmentResponse> completeAppointment(
        @PathVariable final Integer appointmentId
    ) {
        final Appointment appointment = appointmentService.complete(appointmentId);
        return ResponseEntity.status(OK).body(appointmentMapper.toResponse(appointment));
    }

    @ResponseStatus(NO_CONTENT)
    @Operation(
        summary = "Delete appointment",
        description = "Deletes an appointment from the system"
    )
    @DeleteMapping(path = APPOINTMENT_BY_ID)
    public ResponseEntity<Void> deleteAppointment(
        @PathVariable final Integer appointmentId
    ) {
        appointmentService.delete(appointmentId);
        return ResponseEntity.status(NO_CONTENT).build();
    }
}

File: client/src/lib/api/appointment/AppointmentController.ts
/**
 * Appointment API Controller
 *
 * Thin API layer for appointment-related operations.
 * Handles HTTP communication only - no business logic.
 */
import { client } from '$lib/api/client';
import type {
	AppointmentResponse,
	CreateAppointmentRequest,
	UpdateAppointmentRequest
} from '$lib/api/appointment/types';

/**
 * Get all appointments
 */
export async function getAppointments(): Promise<AppointmentResponse[]> {
	const { data, error } = await client.GET('/v1/appointments');
	if (error) throw error;
	return data ?? [];
}

/**
 * Get appointment by ID
 */
export async function getAppointmentById(appointmentId: number): Promise<AppointmentResponse> {
	const { data, error } = await client.GET('/v1/appointments/{appointmentId}', {
		params: { path: { appointmentId } }
	});
	if (error) throw error;
	if (!data) throw new Error('Appointment not found');
	return data;
}

/**
 * Create a new appointment
 */
export async function createAppointment(request: CreateAppointmentRequest): Promise<AppointmentResponse> {
	const { data, error } = await client.POST('/v1/appointments', {
		body: request
	});
	if (error) throw error;
	if (!data) throw new Error('Failed to create appointment');
	return data;
}

/**
 * Update an existing appointment
 */
export async function updateAppointment(
	appointmentId: number,
	request: UpdateAppointmentRequest
): Promise<AppointmentResponse> {
	const { data, error } = await client.PUT('/v1/appointments/{appointmentId}', {
		params: { path: { appointmentId } },
		body: request
	});
	if (error) throw error;
	if (!data) throw new Error('Failed to update appointment');
	return data;
}

/**
 * Cancel an appointment
 */
export async function cancelAppointment(appointmentId: number): Promise<AppointmentResponse> {
	const { data, error } = await client.PATCH('/v1/appointments/{appointmentId}/cancel', {
		params: { path: { appointmentId } }
	});
	if (error) throw error;
	if (!data) throw new Error('Failed to cancel appointment');
	return data;
}

/**
 * Complete an appointment
 */
export async function completeAppointment(appointmentId: number): Promise<AppointmentResponse> {
	const { data, error } = await client.PATCH('/v1/appointments/{appointmentId}/complete', {
		params: { path: { appointmentId } }
	});
	if (error) throw error;
	if (!data) throw new Error('Failed to complete appointment');
	return data;
}

/**
 * Delete an appointment
 */
export async function deleteAppointment(appointmentId: number): Promise<void> {
	const { error } = await client.DELETE('/v1/appointments/{appointmentId}', {
		params: { path: { appointmentId } }
	});
	if (error) throw error;
}
